runtime/src/utils/handle.rs

```diff
@@ -198,6 +204,59 @@ impl MetricHandle {
     }
 }
 
+/// A type alias for a panic.
+pub(crate) type Panic = Box<dyn Any + Send + 'static>;
+
+/// Notifies the runtime when a spawned task panics, so it can propagate the failure.
+#[derive(Clone)]
+pub(crate) struct Panicker {
+    sender: Arc<Mutex<Option<oneshot::Sender<Panic>>>>,
+}
+
+impl Panicker {
+    /// Creates a new [Panicker] with the provided sender.
+    pub(crate) fn new(sender: oneshot::Sender<Panic>) -> Self {
+        Self {
+            sender: Arc::new(Mutex::new(Some(sender))),
+        }
+    }
+
+    /// Notifies the [Panicker] that a panic has occurred.
+    pub(crate) fn notify(&self, panic: Panic) {
+        if let Some(sender) = self.sender.lock().unwrap().take() {
+            let _ = sender.send(panic);
+        }
+    }
+
+    /// Polls a task that should be interrupted by a panic.
+    pub(crate) async fn interrupt<Fut>(
+        panicked: Option<oneshot::Receiver<Panic>>,
+        task: Fut,
+    ) -> Fut::Output
+    where
+        Fut: Future,
+    {
+        // Wait for task to complete (if we are catching panics)
+        let Some(panicked) = panicked else {
+            return task.await;
+        };
+
+        // Wait for task to complete or panic
+        pin_mut!(panicked);
+        pin_mut!(task);
+        match select(panicked, task).await {
+            Either::Left((panic, task)) => match panic {
+                // If the oneshot completes, resume the unwind
+                Ok(panic) => resume_unwind(panic),
+                // If the oneshot closes, wait for the task to complete
+                Err(_) => task.await,
```

Comment:
If there is no panic, we should continue polling the task (probably some odd shutdown case).